core/node/libp2p.go

```diff
@@ -0,0 +1,579 @@
+	"github.com/libp2p/go-libp2p-routing"
+	"github.com/libp2p/go-libp2p-routing-helpers"
+	p2pbhost "github.com/libp2p/go-libp2p/p2p/host/basic"
+	"github.com/libp2p/go-libp2p/p2p/host/routed"
+	mafilter "github.com/libp2p/go-maddr-filter"
+	smux "github.com/libp2p/go-stream-muxer"
+	ma "github.com/multiformats/go-multiaddr"
+	mplex "github.com/whyrusleeping/go-smux-multiplex"
+	yamux "github.com/whyrusleeping/go-smux-yamux"
+	"github.com/whyrusleeping/multiaddr-filter"
+	mamask "github.com/whyrusleeping/multiaddr-filter"
+	"go.uber.org/fx"
+
+	"github.com/ipfs/go-ipfs/repo"
+)
+
+var log = logging.Logger("node")
+
+type HostOption func(ctx context.Context, id peer.ID, ps peerstore.Peerstore, options ...libp2p.Option) (host.Host, error)
+type RoutingOption func(context.Context, host.Host, datastore.Batching, record.Validator) (routing.IpfsRouting, error)
+
+var DefaultHostOption HostOption = constructPeerHost
+
+// isolates the complex initialization steps
+func constructPeerHost(ctx context.Context, id peer.ID, ps peerstore.Peerstore, options ...libp2p.Option) (host.Host, error) {
+	pkey := ps.PrivKey(id)
+	if pkey == nil {
+		return nil, fmt.Errorf("missing private key for node ID: %s", id.Pretty())
+	}
+	options = append([]libp2p.Option{libp2p.Identity(pkey), libp2p.Peerstore(ps)}, options...)
+	return libp2p.New(ctx, options...)
+}
+
+func constructDHTRouting(ctx context.Context, host host.Host, dstore datastore.Batching, validator record.Validator) (routing.IpfsRouting, error) {
+	return dht.New(
+		ctx, host,
+		dhtopts.Datastore(dstore),
+		dhtopts.Validator(validator),
+	)
+}
+
+func constructClientDHTRouting(ctx context.Context, host host.Host, dstore datastore.Batching, validator record.Validator) (routing.IpfsRouting, error) {
+	return dht.New(
+		ctx, host,
+		dhtopts.Client(true),
+		dhtopts.Datastore(dstore),
+		dhtopts.Validator(validator),
+	)
+}
+
+var DHTOption RoutingOption = constructDHTRouting
+var DHTClientOption RoutingOption = constructClientDHTRouting
+var NilRouterOption RoutingOption = nilrouting.ConstructNilRouting
+
+func Peerstore(id peer.ID, sk crypto.PrivKey) peerstore.Peerstore {
+	ps := pstoremem.NewPeerstore()
+
+	if sk != nil {
+		ps.AddPrivKey(id, sk)
+		ps.AddPubKey(id, sk.GetPublic())
+	}
+
+	return ps
+}
+
+func P2PAddrFilters(cfg *config.Config) (opts Libp2pOpts, err error) {
+	for _, s := range cfg.Swarm.AddrFilters {
+		f, err := mask.NewMask(s)
+		if err != nil {
+			return opts, fmt.Errorf("incorrectly formatted address filter in config: %s", s)
+		}
+		opts.Opts = append(opts.Opts, libp2p.FilterAddresses(f))
+	}
+	return opts, nil
+}
+
+func P2PBandwidthCounter(cfg *config.Config) (opts Libp2pOpts, reporter metrics.Reporter) {
+	reporter = metrics.NewBandwidthCounter()
+
+	if !cfg.Swarm.DisableBandwidthMetrics {
+		opts.Opts = append(opts.Opts, libp2p.BandwidthReporter(reporter))
+	}
+	return opts, reporter
+}
+
+type Libp2pOpts struct {
+	fx.Out
+
+	Opts []libp2p.Option `group:"libp2p"`
+}
```

Comment:
We should be able to get rid of this helper entirely (https://github.com/libp2p/go-libp2p/pull/603). I'll file a followup PR after merging this.